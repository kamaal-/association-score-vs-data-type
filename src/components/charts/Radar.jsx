import React, {useRef, useEffect} from 'react'
import { select, range, scaleLinear} from 'd3'
import { getRadian, getScalableMethod, getX1Angle, getY1Angle, setSvgLines } from './helper'
import {COLORS} from '../../config/'

function Radar({row, title}) {
    const svgRef = useRef()
    const wrapperRef = useRef()
    let svg = null // to bind d3 svg el
    const marginTop = 30

    useEffect(() => {
        const $el = svgRef.current
        const $wrapper = wrapperRef.current
        if($el && $wrapper){
            const originalHeight = $wrapper.offsetHeight
            const originalWidth = $wrapper.offsetWidth
            const radarHeight = originalHeight - 50;
            // eslint-disable-next-line react-hooks/exhaustive-deps
            svg = select($el)
            const data = row.visualizerData
            // axis
            const radius = radarHeight/2;
            const steps = range(5)
            const scaleSteps = scaleLinear().range([0,1]).domain([0,4])
            const axisAngle = getScalableMethod(scaleLinear, [0, 7], [0, getRadian(2)]);
            const measureScale = getScalableMethod(scaleLinear, [0, 1], [0, radius]);
            const stepScale = getScalableMethod(scaleLinear, [0, 4], [0, radius]);

            // create svg group to hold radar axis
            const svgWrapper = svg
                .append('g')
                .attr('transform', `translate(${originalWidth/2}, ${(radarHeight/2) + marginTop}) rotate(-90)`)

            setSvgLines(svgWrapper, data, 'lines', COLORS.lineColors, '1', getX1Angle.bind(null, axisAngle, radius), getY1Angle.bind(null, axisAngle, radius))
            
            const textX = i => (Math.cos(axisAngle(i)) * radius)
            const textY = i => (Math.sin(axisAngle(i)) * radius)
            const RADIAN_OFFSET = Math.PI/2;
            
            const labelAnchor = (d, i) => {
                const topAngle = 0 - Math.PI/2;
                const bottomAngle = 0 + Math.PI/2;
                const the_angle = axisAngle(i) - RADIAN_OFFSET;
                let anchor = "";
                if (the_angle === topAngle || the_angle === bottomAngle)
                  anchor = "middle"
                else if(the_angle > topAngle && the_angle < bottomAngle)
                  anchor = "start"
                else
                  anchor = "end"
                return anchor;
            }

            svgWrapper.selectAll('axis_labels')
                .data(data)
                .enter()
                .append('g')
                .attr("transform", (_, i) => {
                    const x = textX(i)
                    const y = textY(i)
                    const topAngle = 0 - Math.PI/2;
                    const bottomAngle = 0 + Math.PI/2;
                    const the_angle = axisAngle(i) - RADIAN_OFFSET;
                    const d = {
                        dx: -7,
                        dy: -10
                    }
                    if (the_angle === topAngle || the_angle === bottomAngle){
                        d.dx = 10
                        d.dy = 10
                    }
                    if(the_angle > topAngle && the_angle < bottomAngle){
                        d.dx = -7
                        d.dy = 10
                    }
                    return `translate(${x+y + d.dx},${-x + y + d.dy}) rotate(90)`
                })
                .append('text')
                .attr('x', (d,i) => (Math.cos(axisAngle(i)) * radius))
                .style('text-anchor', labelAnchor)
                .attr('fill', COLORS.primaryTextColor)
                .attr('data-testid', 'radar-axis-labels')
                .style("font-size",12)
                .attr('y', (d,i) => (Math.sin(axisAngle(i)) * radius))
                .text((d) => d.label)

            const polygonWrapper = svgWrapper.append('g')
            const stepTextWrapper = svgWrapper.append('g').attr('transform', 'rotate(90)')

            // adding rings
            steps.forEach(s => {
                const rad = stepScale(s);

                // rings
                polygonWrapper.selectAll('polygon__ring')
                    .data([data]).enter()
                    .append('polygon')
                    .classed('ring',true)
                    .attr('fill', 'transparent')
                    .attr('stroke', COLORS.lineColors)
                    .attr('data-testid', 'radar-steps-poly')
                    .attr('points', function(d) {
                    return d.map(function(d,i) {
                        return [Math.cos(axisAngle(i)) * rad, Math.sin(axisAngle(i)) * rad];
                    }).join(' ');
                    })
                
                // ring value labels
                stepTextWrapper
                    .append('text')
                    .attr('y', () =>  (-stepScale(s) + 5))
                    .attr('x', 5)
                    .attr('data-testid', 'radar-steps-text')
                    .text(() => scaleSteps(s))
                    .attr('fill', COLORS.primaryTextColor)
                    .style("font-size",12)
            })

            // plot
            svgWrapper.selectAll('polygon__area')
                .data([data]).enter()
                .append('polygon')
                .classed('area',true)
                .attr('fill', 'rgba(65, 41, 96, 0.7)')
                .attr('stroke', 'transparent')
                .attr('stroke-width', '0')
                .attr('stroke-linejoin', "bevel" )
                .attr('data-testid', 'radar-poly')
                .attr('points', function(d) {
                  return d.map(function(d,i) {
                    const rad = measureScale(d.value);
                    return [Math.cos(axisAngle(i)) * rad, Math.sin(axisAngle(i)) * rad];
                  }).join(' ');
                })
        }
    }, [])

    return (
        <div style={{height: '100%'}} ref={wrapperRef}>
            <h4 style={{margin: 0, height: '20px', marginBottom: '10px'}}>{title}</h4>
            <svg style={{width: '100%', height: `100%`}} ref={svgRef}/>
        </div>
    )
}

export default Radar