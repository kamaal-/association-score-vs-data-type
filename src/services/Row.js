import {initialDataTypes} from '../config/'

export default class Row {
    constructor(data){
        this.data = data 
        this.setupDataTypes()
    }

    setupDataTypes(data){
        const self = this
        this.dataTypeKeyValues = (Object.keys(initialDataTypes)|| []).map( key => {
            return {...initialDataTypes[key], value: self.data.association_score.datatypes[key]}
        })
    }

    get geneId(){
        return this.data.target.id
    }

    get symbol(){
        return this.data.target.gene_info.symbol
    }

    get name(){
        return this.data.target.gene_info.name
    }

    get overallScore(){
        return this.data.association_score.overall
    }

    get visualizerData(){
        return this.dataTypeKeyValues
    }

    set dataType(dataType){
        this.dataTypeKeyValues.push(dataType)
    }
}