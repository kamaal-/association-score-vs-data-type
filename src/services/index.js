import Axios from "axios"
import Row from './Row'
import {totalItemsFromTo} from '../config/'

export const dataToRow = d => new Row(d)

export const transformDataToRows  = data => ((data && data.data) || [])
    .map(dataToRow)
    .sort(sortRowByOverallScore)
    .slice(...totalItemsFromTo)

export const sortRowByOverallScore = (a, b) => {
    if(a.overallScore > b.overallScore) return -1;
    if(a.overallScore < b.overallScore) return 1;
    return 0;
}

export const getInitialData = (cb) => {
    Axios.get('https://demo6922545.mockable.io/')
        .then(({data}) => cb(transformDataToRows (data)))
        .catch(e => {
            console.log(e)
        })
}