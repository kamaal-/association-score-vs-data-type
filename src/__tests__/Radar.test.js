import React from 'react';
import { render } from '@testing-library/react';
import Radar from '../components/charts/Radar';
import data from '../config/mock.data';
import { transformDataToRows Â } from '../services/'
import ReactDOM from 'react-dom'

describe('Radar Chart', () => {

    const setup = () => {
        const container = document.createElement('div')
        const rows = transformDataToRows(data())	
        return render(<Radar row={rows[0]} chartTitle={'Chart Title'}/>, {
            container: document.body.appendChild(container)
        });
    }

  	it('renders without crashing', () => {
        const div = document.createElement('div');
        const rows = transformDataToRows(data())	
    	ReactDOM.render(<Radar row={rows[0]} chartTitle={'Chart Title'}/>, div);
    	ReactDOM.unmountComponentAtNode(div);
    });

  	it('Should polygon', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('radar-poly').length).toBe(1);
    });
      
    it('Should render 5 steps text', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('radar-steps-text').length).toBe(5);
    });
      
    it('Should render 5 steps polygon', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('radar-steps-poly').length).toBe(5);
  	});//radar-axis-labels

    it('Should render 7 axis labels', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('radar-axis-labels').length).toBe(7);
      });//axis-lines
      
    it('Should render 7 axis lines', () => {
		const { getAllByTestId } = setup()
		expect( getAllByTestId('axis-lines').length).toBe(7);
    });
      
    it('should match snapshot', () => {
        const { asFragment } = setup()
        expect(asFragment()).toMatchSnapshot();
    })

})
 