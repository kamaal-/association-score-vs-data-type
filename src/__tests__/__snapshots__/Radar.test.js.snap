// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radar Chart should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="height: 100%;"
  >
    <h4
      style="margin: 0px 0px 10px 0px; height: 20px;"
    />
    <svg
      style="width: 100%; height: 100%;"
    >
      <g
        transform="translate(0, 5) rotate(-90)"
      >
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="-25"
          y0="0"
          y1="0"
        />
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="-15.58724504646834"
          y0="0"
          y1="-19.545787061700747"
        />
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="5.563023348907858"
          y0="0"
          y1="-24.37319780454559"
        />
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="22.524221697560478"
          y0="0"
          y1="-10.847093477938955"
        />
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="22.524221697560478"
          y0="0"
          y1="10.84709347793895"
        />
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="5.563023348907865"
          y0="0"
          y1="24.37319780454559"
        />
        <line
          class="lines"
          data-testid="axis-lines"
          stroke="rgba(65, 41, 96, 0.3)"
          stroke-width="1"
          x0="0"
          x1="-15.587245046468334"
          y0="0"
          y1="19.545787061700747"
        />
        <g
          transform="translate(-15,35) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: middle; font-size: 12;"
            x="-25"
            y="0"
          >
            Literature
          </text>
        </g>
        <g
          transform="translate(-42.13303210816909,6.041457984767593) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: start; font-size: 12;"
            x="-15.58724504646834"
            y="-19.545787061700747"
          >
            RNA Expression
          </text>
        </g>
        <g
          transform="translate(-25.810174455637732,-19.93622115345345) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: start; font-size: 12;"
            x="5.563023348907858"
            y="-24.37319780454559"
          >
            Genetic Association
          </text>
        </g>
        <g
          transform="translate(4.6771282196215225,-23.371315175499433) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: start; font-size: 12;"
            x="22.524221697560478"
            y="-10.847093477938955"
          >
            Somatic Mutation
          </text>
        </g>
        <g
          transform="translate(26.371315175499426,-21.67712821962153) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: end; font-size: 12;"
            x="22.524221697560478"
            y="10.84709347793895"
          >
            Known Drug
          </text>
        </g>
        <g
          transform="translate(22.936221153453456,8.810174455637725) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: end; font-size: 12;"
            x="5.563023348907865"
            y="24.37319780454559"
          >
            Animal Model
          </text>
        </g>
        <g
          transform="translate(-3.0414579847675878,25.133032108169083) rotate(90)"
        >
          <text
            data-testid="radar-axis-labels"
            fill="#412960"
            style="text-anchor: end; font-size: 12;"
            x="-15.587245046468334"
            y="19.545787061700747"
          >
            Affected Pathway
          </text>
        </g>
        <g>
          <polygon
            class="ring"
            data-testid="radar-steps-poly"
            fill="transparent"
            points="0,0 0,0 0,0 0,0 0,0 0,0 0,0"
            stroke="rgba(65, 41, 96, 0.3)"
          />
          <polygon
            class="ring"
            data-testid="radar-steps-poly"
            fill="transparent"
            points="-6.25,0 -3.896811261617085,-4.886446765425187 1.3907558372269646,-6.093299451136398 5.631055424390119,-2.711773369484739 5.631055424390119,2.7117733694847375 1.3907558372269662,6.093299451136398 -3.8968112616170836,4.886446765425187"
            stroke="rgba(65, 41, 96, 0.3)"
          />
          <polygon
            class="ring"
            data-testid="radar-steps-poly"
            fill="transparent"
            points="-12.5,0 -7.79362252323417,-9.772893530850373 2.781511674453929,-12.186598902272795 11.262110848780239,-5.423546738969478 11.262110848780239,5.423546738969475 2.7815116744539323,12.186598902272795 -7.793622523234167,9.772893530850373"
            stroke="rgba(65, 41, 96, 0.3)"
          />
          <polygon
            class="ring"
            data-testid="radar-steps-poly"
            fill="transparent"
            points="-18.75,0 -11.690433784851255,-14.659340296275559 4.172267511680894,-18.279898353409195 16.89316627317036,-8.135320108454216 16.89316627317036,8.135320108454213 4.172267511680898,18.279898353409195 -11.690433784851251,14.65934029627556"
            stroke="rgba(65, 41, 96, 0.3)"
          />
          <polygon
            class="ring"
            data-testid="radar-steps-poly"
            fill="transparent"
            points="-25,0 -15.58724504646834,-19.545787061700747 5.563023348907858,-24.37319780454559 22.524221697560478,-10.847093477938955 22.524221697560478,10.84709347793895 5.563023348907865,24.37319780454559 -15.587245046468334,19.545787061700747"
            stroke="rgba(65, 41, 96, 0.3)"
          />
        </g>
        <g
          transform="rotate(90)"
        >
          <text
            data-testid="radar-steps-text"
            fill="#412960"
            style="font-size: 12;"
            x="5"
            y="5"
          >
            0
          </text>
          <text
            data-testid="radar-steps-text"
            fill="#412960"
            style="font-size: 12;"
            x="5"
            y="11.25"
          >
            0.25
          </text>
          <text
            data-testid="radar-steps-text"
            fill="#412960"
            style="font-size: 12;"
            x="5"
            y="17.5"
          >
            0.5
          </text>
          <text
            data-testid="radar-steps-text"
            fill="#412960"
            style="font-size: 12;"
            x="5"
            y="23.75"
          >
            0.75
          </text>
          <text
            data-testid="radar-steps-text"
            fill="#412960"
            style="font-size: 12;"
            x="5"
            y="30"
          >
            1
          </text>
        </g>
        <polygon
          class="area"
          data-testid="radar-poly"
          fill="rgba(65, 41, 96, 0.7)"
          points="-8.158835581047269,0 -1.0125163936859527,-1.2696553989154602 5.563023348907858,-24.37319780454559 22.524221697560478,-10.847093477938955 22.524221697560478,10.84709347793895 0,0 -15.587245046468334,19.545787061700747"
          stroke="transparent"
          stroke-linejoin="bevel"
          stroke-width="0"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;
